protocols/gossipsub/src/behaviour.rs

```diff
@@ -0,0 +1,1165 @@
+            // TODO: Check if the random sequence numbers causes issues with other clients.
+            // To be interoperable with the go-implementation this is treated as a 64-bit
+            // big-endian uint.
+            sequence_number: rand::random::<[u8; 8]>().to_vec(),
+            topics: topic.into_iter().map(|t| self.topic_hash(t)).collect(),
+        };
+
+        debug!("Publishing message: {:?}", message.id());
+
+        // forward the message to mesh and floodsub peers
+        let local_peer_id = self.local_peer_id.clone();
+        self.forward_msg(message.clone(), &local_peer_id);
+
+        let mut recipient_peers = HashMap::new();
+        for topic_hash in &message.topics {
+            // if not subscribed to the topic, use fanout peers
+            if self.mesh.get(&topic_hash).is_none() {
+                debug!("Topic: {:?} not in the mesh", topic_hash);
+                // build a list of peers to forward the message to
+                // if we have fanout peers add them to the map
+                if self.fanout.contains_key(&topic_hash) {
+                    for peer in self.fanout.get(&topic_hash).expect("Topic must exist") {
+                        recipient_peers.insert(peer.clone(), ());
+                    }
+                } else {
+                    // we have no fanout peers, select mesh_n of them and add them to the fanout
+                    let mesh_n = self.config.mesh_n;
+                    let new_peers = self.get_random_peers(&topic_hash, mesh_n, { |_| true });
+                    // add the new peers to the fanout and recipient peers
+                    self.fanout.insert(topic_hash.clone(), new_peers.clone());
+                    for peer in new_peers {
+                        debug!("Peer added to fanout: {:?}", peer);
+                        recipient_peers.insert(peer.clone(), ());
+                    }
+                }
+                // we are publishing to fanout peers - update the time we published
+                self.fanout_last_pub
+                    .insert(topic_hash.clone(), Instant::now());
+            }
+        }
+
+        // add published message to our received caches
+        self.mcache.put(message.clone());
+        self.received.put(message.id(), ());
+
+        let event = Arc::new(GossipsubRpc {
+            subscriptions: Vec::new(),
+            messages: vec![message.clone()],
+            control_msgs: Vec::new(),
+        });
+        // Send to peers we know are subscribed to the topic.
+        for peer_id in recipient_peers.keys() {
+            debug!("Sending message to peer: {:?}", peer_id);
+            self.events.push_back(NetworkBehaviourAction::SendEvent {
+                peer_id: peer_id.clone(),
+                event: event.clone(),
+            });
+        }
+        info!("Published message: {:?}", message.id());
+    }
+
+    /// This function should be called when `config.manual_propagation` is `true` to order to
```

Comment:
```suggestion
    /// This function should be called when `config.manual_propagation` is `true` in order to
```

Not a native speaker, just a guess.

Reply:
Correct